### PR TITLE
Better assignability errors to never-producing vacuous intersections

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This was a naive first pass at something that I realized would be nice as I wrote up the breaking change section.

I couldn't do a fix everywhere because it became pretty impractical - for example, by the time we do the checks in some places, assignability, I guess we're given `never` instead of the original intersection.

```ts
const x1 = { a: 'foo', b: 42 };
const x2 = { a: 'foo', b: true };

declare let k: 'a' | 'b';

x1[k] = 42;
```

Things I'd like some guidance on:

1. Uh, I just hit `1 << 29` on `NodeBuilderFlags`. I've forgotten if that means I get a prize or a scolding for hitting the max unboxed value on a flag. (thanks @yuit) image
1. Error message phrasing.

### CODE REVIEW

File: tests/baselines/reference/intersectionReduction.errors.txt [hunk 1]
```diff
@@ -44,6 +47,11 @@ tests/cases/conformance/types/intersection/intersectionReduction.ts(81,1): error
     ab.kind;  // Error
        ~~~~
 !!! error TS2339: Property 'kind' does not exist on type 'never'.
+!!! error TS2339:   The type 'never' was reduced from the intersection 'A & B'. Each type of that intersection has properties that conflict, so values of that type can never exist.
```
Comment (weswigham) [inline tests/baselines/reference/intersectionReduction.errors.txt]: You should probably search for and report the properties that conflict and their respective types, no? So the error is somewhat more actionable?

Comment (DanielRosenwasser) [reply tests/baselines/reference/intersectionReduction.errors.txt]: Definitely seems doable, but only if the approach in this PR seems reasonable.

Comment (weswigham) [reply tests/baselines/reference/intersectionReduction.errors.txt]: Seems OK. We could probably consider leaving behind more "breadcrumbs" like this in the future to explain how a specific type came about; but it's always about striking that balance between useful explaianation and terseness.

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -5653,5 +5653,9 @@
     "An optional chain cannot contain private identifiers.": {
         "category": "Error",
         "code": 18030
+    },
+    "The type 'never' was reduced from the intersection '{0}'. Each type of that intersection has properties that conflict, so values of that type can never exist.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:5657]: why 'each' here? Wouldn't 'some' be correct?

If that's true, I'd just say "Types in that intersection ..."

**Review by sandersn — COMMENTED** [review]
Error seems good, and implementation seems OK except for the flags problem.

File: src/compiler/diagnosticMessages.json (hunk 2 above)
Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:5657]: Good point

**Comment by DanielRosenwasser** [comment]
#37618 provides the better error message.

**Pull request closed without merging**